plugins/Login/Security/BruteForceDetection.php

```diff
@@ -0,0 +1,134 @@
+<?php
+/**
+ * Matomo - free/libre analytics platform
+ *
+ * @link https://matomo.org
+ * @license http://www.gnu.org/licenses/gpl-3.0.html GPL v3 or later
+ *
+ */
+namespace Piwik\Plugins\Login\Security;
+
+use Piwik\Common;
+use Piwik\Date;
+use Piwik\Db;
+use Piwik\Plugins\Login\SystemSettings;
+
+class BruteForceDetection {
+
+    private $minutesTimeRange;
+    private $maxLogAttempts;
+
+    private $table = 'brute_force_log';
+    private $tablePrefixed = '';
+
+    /**
+     * @var SystemSettings
+     */
+    private $settings;
+
+    public function __construct(SystemSettings $systemSettings)
+    {
+        $this->tablePrefixed = Common::prefixTable($this->table);
+        $this->settings = $systemSettings;
+        $this->minutesTimeRange = $systemSettings->loginAttemptsTimeRange->getValue();
+        $this->maxLogAttempts = $systemSettings->maxFailedLoginsPerMinutes->getValue();
+    }
+
+    public function isEnabled()
+    {
+        return $this->settings->enableBruteForceDetection->getValue();
+    }
+
+    public function addFailedLoginAttempt($ipAddress)
+    {
+        $now = $this->getNow()->getDatetime();
+        $db = Db::get();
+        $db->query('INSERT INTO '.$this->tablePrefixed.' SET ip_address = ?, attempted_at = ?', array($ipAddress, $now));
+    }
+
+    public function canLogin($ipAddress)
+    {
+        if ($this->settings->isBlacklistedIp($ipAddress)) {
+            return false;
+        }
+
+        if ($this->settings->isWhitelistedIp($ipAddress)) {
+            return true;
+        }
+
+        $db = Db::get();
+
+        $startTime = $this->getStartTimeRange();
+        $sql = 'SELECT count(*) as numLogins FROM '.$this->tablePrefixed.' WHERE ip_address = ? AND attempted_at > ?';
+        $numLogins = $db->fetchOne($sql, array($ipAddress, $startTime));
+
+        return empty($numLogins) || $numLogins <= $this->maxLogAttempts;
+    }
+
+    public function getCurrentlyBlockedIps()
+    {
+        $sql = 'SELECT ip_address
+                FROM ' . $this->tablePrefixed . ' 
+                WHERE attempted_at > ?
+                GROUP BY ip_address 
+                HAVING count(*) > ' . (int) $this->maxLogAttempts;
```

Comment:
This query will do a full table scan, correct? Since there's no index on attempted_at. Would that be an issue for tracking (assuming there are enough roes in the table)?

Reply:
I figured because we delete entries after 7 days there shouldn't be too many in the DB and it should be fine and the query is only executed in a task or when viewing blocked IPs.

Reply:
👍